File: code/cpp/container-with-most-water.cc
// 19 ms, 09/18/2017
// Time  : O(n)
// Space : O(1)
// TODO  : Take a look at the solution below
class Solution {
public:
    int maxArea(vector<int>& height)
    {
        int max_area = 0;
        for (int i = 0, j = height.size() - 1; i < j; ) {
            max_area = max(max_area, (j - i) * min(height[i], height[j]));
            if (height[i] > height[j])
                --j;
            else if (height[i] < height[j])
                ++i;
            else
                ++i, --j;
        }
        return max_area;
    }
};
/*--------------------------------------------------------*/
// A more compact solution on leetcode
class Solution {
public:
    int maxArea(vector<int>& height) {
        int water = 0;
        int i = 0, j = height.size() - 1;
        while (i < j){
            int h = min(height[i], height[j]);
            water = max(water, (j - i) * h);
            while (height[i] <= h && i < j) i++;
            while (height[j] <= h && i < j) j--;
        }
        return water;
    }
};

File: code/python/container-with-most-water.py
# 01/12/2018, 108 ms
# Time  : O(n)
# Space : O(1)
class Solution:
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        # Use two-pointer technique - start at the extreme ends and
        # shrink the range on the side that's shorter in the hope of
        # getting a taller bar, thereby potentially getting a larger
        # area/volume
        A = height
        left, right = 0, len(A) - 1
        res = 0
        while left < right:
            res = max(res, (right - left) * min(A[right], A[left]))
            if A[left] < A[right]:
                left += 1
            else:
                right -= 1
        return res
